### ISSUE TITLE
How to reset trained adapter model?

### ISSUE THREAD

**Issue by DavidPfl** [Adapter-Hub/adapter-transformers#179]
## Environment info

     
- `adapter-transformers` version:
- Platform:
- Python version:
- PyTorch version (GPU?):

## Details

Hi all,

I am trying to train a model with an adapter in a loop.
In each iteration, I want the model to be initialized and train the adapter from scratch (only the adapter, no full finetuning).
My question is: If I initialize the model as shown in the snippet below, is the model initialized with random weights or does the training in each iteration start with weights from the previous iteration?
Here is the snippet for my question:

        model =  AutoModelWithHeads.from_pretrained(
            model_name,
            config=config,
        )
        try:
            # Add a new adapter --> This raises a ValueError if the adapter was already added in a previous iteration
            model.add_adapter("test")
            # Add a matching classification head
            model.add_classification_head(
                "test",
                num_labels=2,
                id2label={ 0: "no_test", 1: "test"}
            )
        
        except ValueError:
            # Activate the adapter if it was already there.
            model.set_active_adapter("test")
       
        model.train_adapter("test")

        training_args = TrainingArguments(...)
        trainer = Trainer(...)
        trainer.train()

**Comment by calpt** [comment]
Hi @DavidPfl,

> If I initialize the model as shown in the snippet below, is the model initialized with random weights or does the training in each iteration start with weights from the previous iteration?

Currently, your snippet reuses the weights from the previous iteration as an existing adapter cannot be overwritten (hence the ValueError). We can consider adding an option to overwrite existing adapters.

**Comment by DavidPfl** [comment]
Hi @calpt,

since my initial question, I adjusted the code to use the `model_init` argument in the Trainer initialization. I did so because I read [here]( https://discuss.huggingface.co/t/fixing-the-random-seed-in-the-trainer-does-not-produce-the-same-results-across-runs/3442) that this ensures reproducibility when fixing the random seed.
The loop would then look like this:

```python    
def get_model():
  model =  AutoModelWithHeads.from_pretrained(model_name, config=config)
  try:
      # Add a new adapter
      model.add_adapter("hyperpartisan")
  except ValueError:
      # ValueError  is raised when adapter already exists.
      pass
  try:
      model.add_classification_head(
          "hyperpartisan",
          num_labels=2,
          id2label={ 0: "neutral", 1: "hyperpartisan"}
      )
  except ValueError:
      # ValueError  is raised when adapter already exists.
      pass
  finally:
      model.set_active_adapters("hyperpartisan")
      model.train_adapter("hyperpartisan")
  return model

for _ in range(2):
  # here I change some hyperparameters for grid search
  training_args = TrainingArguments(...)
  trainer = Trainer(model_init = get_model, ...)
  trainer.train()
```

Would this initialize the adapter with the same random weights in each iteration or would I continue with the same weights as from last iteration?
If the weights are still not reinitialized, how can I overwrite the existing adapter or make sure that in each loop iteration I start the adapter weights from scratch?
Maybe I am trying to fix an unneccesary problem. My intention is to do a grid search over some hyperparameters in a for-loop.
Is there a better way for doing this?

**Comment by calpt** [comment]
Hi @DavidPfl,

> Would this initialize the adapter with the same random weights in each iteration or would I continue with the same weights as from last iteration?

It would reuse the weights.

We figured that there currently is no straightforward to do what you are describing. So, based on your use case, we're adding a `delete_adapter()` method in the next version. There will also be a new parameter `overwrite_ok` for `add_adapter()` to overwrite the weights each time `add_adapter()` is called. See https://github.com/Adapter-Hub/adapter-transformers/pull/189.

Hope this helps!

**Comment by calpt** [comment]
> So, based on your use case, we're adding a delete_adapter() method in the next version. There will also be a new parameter overwrite_ok for add_adapter() to overwrite the weights each time add_adapter() is called.

This was released with v2.1.0.
